### PR TITLE
Better handling of standard situations

### PR DESCRIPTION
**Pull request by SammyRamone**
## Proposed changes
Improves handling of standard situations and listening to game controller

## Necessary checks
- [ ] Update package version
- [ ] Run `catkin build`
- [ ] Write documentation
- [ ] Create issues for future work
- [ ] Test on your machine
- [ ] Test on the robot
- [x] Put the PR on our Project board

### CODE REVIEW

File: bitbots_body_behavior/src/bitbots_body_behavior/decisions/allowed_to_move.py [hunk 1]
```diff
@@ -6,29 +6,38 @@
...
             return 'NORMAL'
         elif self.blackboard.gamestate.get_gamestate() == GameState.GAMESTATE_PLAYING:
-            if not self.blackboard.gamestate.has_kickoff() and self.blackboard.gamestate.get_secondary_seconds_remaining() != 0:
+            if not self.blackboard.gamestate.has_kickoff():
+                if self.blackboard.gamestate.get_secondary_seconds_remaining() == 0:
+                    # time is up for the other team
+                    return 'NORMAL'
+                ball_pos = self.blackboard.world_model.get_ball_position_xy()
+                # if we know where the ball is and that it moved, we can play too
+                if rospy.Time.now() - self.blackboard.world_model.ball_last_seen() < self.ball_lost_time and (
```
Comment (Flova) [inline bitbots_body_behavior/src/bitbots_body_behavior/decisions/allowed_to_move.py:36]: Why do we check for a lost ball in the allowed-to-move decision? Normally this is done in the dsd file.

Comment (Flova) [reply bitbots_body_behavior/src/bitbots_body_behavior/decisions/allowed_to_move.py:36]: I would add another decision for the ball moved.

Comment (SammyRamone) [reply bitbots_body_behavior/src/bitbots_body_behavior/decisions/allowed_to_move.py:36]: I would keep it for now like this. I agree that things could be improved, but there are a lot of things all over the body behavior which need to be refactored and I would rather like to do this after the competition.

Comment (SammyRamone) [reply bitbots_body_behavior/src/bitbots_body_behavior/decisions/allowed_to_move.py]: It now publishes additional debug data

File: bitbots_body_behavior/src/bitbots_body_behavior/decisions/closest_to_ball.py [hunk 2]
```diff
@@ -36,3 +36,26 @@ def perform(self, reevaluate=False):
...
+        self.publish_debug_data(f"Rank to ball", rank)
+        if rank == 1:
+            return "FIRST"
+        elif rank == 2:
+            return "SECOND"
+        elif rank == 3:
+            return "THIRD"
+        else:
+            # emergency fall back if something goes wrong
+            return "FIRST"
```
Comment (johagge) [inline bitbots_body_behavior/src/bitbots_body_behavior/decisions/closest_to_ball.py:58]: maybe we should print a warning in this case

File: bitbots_body_behavior/src/bitbots_body_behavior/main.dsd [hunk 3]
```diff
@@ -53,149 +56,121 @@ $TimerRunning + name:reorientation
...
-                            NO --> @ChangeAction + action:positioning, @GoToDefensePosition
-                        NO --> @ChangeAction + action:positioning, @GoToDefensePosition
-                NO --> @Stop, @LookAtFieldFeatures
-        LOW --> $TimerEnded + name:reorientation_pause
-            NO --> #NoLocalizationPlayerBehavior
-            YES --> #ReorientationPause
-    NO --> #NoLocalizationPlayerBehavior
+#StrikerRole
+$GoalieHandlingBall
+    YES --> @ChangeAction + action:positioning, @AvoidBallActive, @GoToPassAcceptPosition //don't go to far back
```
Comment (johagge) [inline bitbots_body_behavior/src/bitbots_body_behavior/main.dsd:82]:
```suggestion
    YES --> @ChangeAction + action:positioning, @AvoidBallActive, @GoToPassAcceptPosition //don't go too far back
```

**Comment by johagge** [comment]
currently the behavior on this branch crashes immediately with the following error:
`INFO] [/body_behavior] [1623941128.452364, 6.432000]: Debugging is active. Publishing on debug/dsd/body_behavior
Traceback (most recent call last):
  File "/home/jonas/catkin_ws/src/bitbots_meta/bitbots_behavior/bitbots_body_behavior/src/bitbots_body_behavior/body_behavior.py", line 45, in <module>
    D.load_behavior(os.path.join(dirname, "main.dsd"))
  File "/home/jonas/catkin_ws/src/bitbots_meta/dynamic_stack_decider/dynamic_stack_decider/src/dynamic_stack_decider/dsd.py", line 137, in load_behavior
    self.tree = parse_dsd(path)
  File "/home/jonas/catkin_ws/src/bitbots_meta/dynamic_stack_decider/dynamic_stack_decider/src/dynamic_stack_decider/parser.py", line 102, in parse
    raise ParseError('Error parsing line {}: {} not defined'.format(lnr, name))
dynamic_stack_decider.parser.ParseError: Error parsing line 31: SearchBall not defined`

File: bitbots_body_behavior/src/bitbots_body_behavior/decisions/closest_to_ball.py (hunk 2 above)
Comment (SammyRamone) [reply bitbots_body_behavior/src/bitbots_body_behavior/decisions/closest_to_ball.py:58]: true

**Review by johagge — APPROVED** [review]

**Pull request merged**